### ISSUE TITLE
Adapting the RoPE for 3D data

### ISSUE THREAD

**Issue by tristanhoefer** [naver-ai/rope-vit#12]
Hey,

First of all, thank you for providing such a great codebase!

I was wondering if it might be possible to adapt your implementation to 3-dimensional data. My initial thought was to extend the frequency components to include a z-axis. Would this approach be mathematically compatible with your method, or could it introduce issues?

I’d greatly appreciate your insights on this.

Best regards,
Tristan

**Comment by bhheo** [comment]
Hi Tristan

I appreciate your interest in our study.

I agree with your initial thought. Since our RoPE-Mixed relies on learnable frequencies, I think simply adding a z-axis might work well for 3D data.
Also, adding the z-axis is valid in the aspect of the 3D Fourier transform.

The most possible issue is the learning feasibility of frequencies. 
In my case, image classification, the frequencies can be trained without changing the optimizer settings, such as learning rate.
But, there is no guarantee that frequencies can be trained with the same optimizer setting in other domains.
So, I recommend trying it first, and if the learning is not stable, change the learning rate of frequencies.
I have never observed instability in learnable frequency, but that cannot be generalized to all domains.

You should change the initialization function. If z frequencies are initialized to zero, the network does not utilize z-axis RoPE.
https://github.com/naver-ai/rope-vit/blob/2e801b6a5cdb4a7a8d95ed555aef5645b58044df/self-attn/rope_self_attn.py#L11-L24

If you are trying video domain, it might be better to consider frame rate and sampling rate for `t_z`. 
But I'm not sure about this.
https://github.com/naver-ai/rope-vit/blob/2e801b6a5cdb4a7a8d95ed555aef5645b58044df/self-attn/rope_self_attn.py#L26-L30

I hope you get good results in your experiments and research.

Best
Heo

**Comment by tristanhoefer** [comment]
Hi Heo,

thank you for the fast and detailed reply. I will then take a deeper look and run some experiments.

Best
Tristan
